python.d/ntpd.chart.py

```diff
@@ -0,0 +1,385 @@
+                             sequence, status, associd, offset, count)
+        return header
+
+
+class System(Base):
+    def __init__(self):
+        self.request = self.get_header()
+
+    def get_data(self, raw):
+        """
+        Extracts key=value pairs with float/integer from ntp response packet data.
+        """
+        data = dict()
+        for key, value in self.regex.findall(raw):
+            data[key] = float(value) * PRECISION
+        return data
+
+
+class Peer(Base):
+    def __init__(self, idx, name):
+        self.id = idx
+        self.real_name = name
+        self.name = name.replace('.', '_')
+        self.request = self.get_header(self.id)
+
+    def get_data(self, raw):
+        """
+        Extracts key=value pairs with float/integer from ntp response packet data.
+        """
+        data = dict()
+        for key, value in self.regex.findall(raw):
+            dimension = '_'.join([self.name, key])
+            data[dimension] = float(value) * PRECISION
+        return data
+
+    def rebuild_request(self, new_idx):
+        self.id = new_idx
+        self.request = self.get_header(self.id)
+
+
+class Service(SocketService):
+    def __init__(self, configuration=None, name=None):
+        SocketService.__init__(self, configuration=configuration, name=name)
+        self.order = list(ORDER)
+        self.definitions = dict(CHARTS)
+        self.definitions.update(PEER_CHARTS)
+
+        self.port = 'ntp'
+        self.dgram_socket = True
+        self.system = System()
+        self.peers = dict()
+        self.request = str()
+        self.retries = 0
+        self.all_peers = self.configuration.get('all_peers', False)
+
+        peer_filter = self.configuration.get('peer_filter', r'127\..*')
+        try:
+            self.peer_filter = re.compile(r'^((0\.0\.0\.0)|({0}))$'.format(peer_filter))
+        except re.error as error:
+            self.error('Compile pattern error : {0}'.format(error))
+            self.peer_filter = None
+
+    def check(self):
+        """
+        Checks if we can get valid systemvars.
+        If not, returns None to disable module.
+        """
+
+        if not self.peer_filter:
+            return None
+
+        self._parse_config()
+
+        self.request = self.system.request
+        raw_systemvars = self._get_raw_data()
+
+        if not self.system.get_data(raw_systemvars):
+            return None
+        return True
+
+    def get_data(self):
+        """
+        Gets systemvars data on each update.
+        Gets peervars data for all peers on each update.
+        """
+        data = dict()
+
+        self.request = self.system.request
+        raw = self._get_raw_data()
+        if not raw:
+            return None
+
+        data.update(self.system.get_data(raw))
+
+        if not self.all_peers:
+            return data
+
+        # TODO: run self.find_new_peers() every N runs?
```

Comment:
@l2isbad I am not sure if we need to do this. This would be required for the case, when a new peer was added. Maybe @Ferroin can give us some input on this? I have no idea if a new peer can appear without a restart. According to @Ferroin it can happen if a pool is configured. If this always means, that another peer was removed at the same time, then we do not need some extra code, because we would get no data from at least one peer.

Reply:
There are a couple of cases where it might matter, namely:

1. If you're using multicast or broadcast client mode (not very common except on big networks with no Windows computers), the total number of peers may change over time because you track everyone who's broadcasting the time.
2. If you're using manycast client mode, the number of peers can also change over time (even less common than broadcast or multicast modes) because you associate with whoever responds.
3. If the server is reconfigured and restarted between collection cycles (for example, moving from one network to another when using DHCP to find NTP servers), the number of peers can change.
4. If you are using a pool of servers like pool.ntp.org, the total number of peers won't change, but who you're talking to as peers may change (and thus you may see new association ID"s added and old ones removed).
5. If the daemon can't resolve a hostname it's configured to use as a server when it goes to do a lookup, that peer gets removed, but it may get re-added if a subsequent lookup succeeds.  Interestingly however, this is not the case for peers who just constantly timeout (they just stay stuck in the .INIT. state).

Reply:
Wow, thank you :) I don't understand most of it but I think that means we should at least periodically check for new peer_ids or just always check it on every update (maybe one more socket connection does not make the difference).

Reply:
Case 4 and case 5 are the ones we would need to worry about.  1 and 2 are not likely to be widely used on server systems (they're too prone to interference, and have accuracy issues), and 3 is not something I would personally expect most netdata users to experience with any regularity.

Overall though, yes, I think it probably should just check on every update (or at least every few updates).